### PR TITLE
Gymkhana scoring plugin

### PR DESCRIPTION
**Pull request by mabelzhang**
Runs station-keeping + navigation scoring plugins at the same time.
Takes the score from station-keeping if the channel has been passed.

Depends on https://github.com/osrf/vrx/pull/224
The goal for the blackbox is specified in Cartesian coordinates, since the pinger YAML is specified in Cartesian.
This places the red marker for station-keeping above the pinger.

These are the topics associated with the scoring plugin.
The `gymkhana` is from the new top-level VORC scoring plugin.
The `gymkhana_blackbox` is from a VRX station-keeping scoring plugin.
The `gymkhana_channel` is from a VRX navigation scoring plugin.
```
$ rostopic list
...
/vorc/gymkhana/debug/contact
/vorc/gymkhana/task/info
/vorc/gymkhana_blackbox/debug/contact
/vorc/gymkhana_blackbox/goal
/vorc/gymkhana_blackbox/pose_error
/vorc/gymkhana_blackbox/rms_error
/vorc/gymkhana_blackbox/task/info
/vorc/gymkhana_channel/debug/contact
/vorc/gymkhana_channel/task/info
...
```

To test:

Keep this command running in a terminal:
```
$ rostopic echo /vorc/gymkhana/task/info
```
Initially, the score is always 0.

Use keyboard / joystick to move the boat into the channel.
Once the boat crosses the last gate (determined by `/vorc/gymkhana_channel/task/info` fields `state == "finished"` and `timed_out == false`), non-zero scores should pop up (taken directly from `score` from `/vorc/gymkhana_blackbox/task/info`).

Essentially, the score is invalid until the boat crosses the last gate.

The only drawback with this invalidation approach is that, if two users do not cross the last gate, but one got farther than the other, they will both get 0.
One possible way we can have their scores be differentiated is to put something different in the `state` field of the top-level plugin Task info topic `/vorc/gymkhana/task/info`. Currently it's `"running"` or `"finished"`. Maybe if we want to give such users non-zero scores, we can indicate in the state whether they crossed the gate? Then that can be used in combination with the raw score to place the user, if we want.

### CODE REVIEW

File: vorc_gazebo/include/vorc_gazebo/gymkhana_scoring_plugin.hh [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  public: virtual ~GymkhanaScoringPlugin();
+
+  // Documentation inherited.
+  public: virtual void Load(gazebo::physics::WorldPtr _world,
+                            sdf::ElementPtr _sdf);
+
+  /// \brief Callback executed at every world update.
+  private: void Update();
+
+  protected: void ChannelCallback(const vrx_gazebo::Task::ConstPtr& msg);
```
Comment (caguero) [inline vorc_gazebo/include/vorc_gazebo/gymkhana_scoring_plugin.hh:43]: Nit: Missing doxygen here and in the next function.

File: vorc_gazebo/launch/gymkhana.launch [hunk 2]
```diff
@@ -8,11 +8,11 @@
   <!-- If true, run gazebo in verbose mode -->
   <arg name="verbose" default="false"/>
   <!-- Set various other gazebo arguments-->
-  <arg name="extra_gazebo_args" default=""/>
+  <arg name="extra_gazebo_args" default="--verbose"/>
```
Comment (caguero) [inline vorc_gazebo/launch/gymkhana.launch]: Is this leftover from testing?

File: vorc_gazebo/launch/gymkhana.launch [hunk 3]
```diff
@@ -8,11 +8,11 @@
...
   <!-- Start in a default namespace -->
   <arg name="namespace" default="cora"/>
   <!-- Do we lock the vessel to the world? -->
-  <arg name="robot_locked" default="false" />
+  <arg name="robot_locked" default="true" />
```
Comment (caguero) [inline vorc_gazebo/launch/gymkhana.launch:15]: Should this be `false`?

File: vorc_gazebo/src/gymkhana_scoring_plugin.cc [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+/////////////////////////////////////////////////
+void GymkhanaScoringPlugin::Update()
+{
+  // Nothing to do if the task is not in "running" state.
+  if (this->ScoringPlugin::TaskState() != "running")
+    return;
+
+  // Check channel navigation is finished
+  // If not, invalidate black box station-keeping score
```
Comment (caguero) [inline vorc_gazebo/src/gymkhana_scoring_plugin.cc:61]: Does this comment apply?

Comment (mabelzhang) [reply vorc_gazebo/src/gymkhana_scoring_plugin.cc:61]: Yes, it's just 0 if channel navigation isn't finished.

File: vorc_gazebo/launch/gymkhana.launch (hunk 3 above)
Comment (mabelzhang) [reply vorc_gazebo/launch/gymkhana.launch:15]: There's an error from station-keeping plugin about unable to unlock the 2 joints, if it's false. I looked at station-keeping and navigation launch files in VRX, and they both locked it, so I just changed it to true.

File: vorc_gazebo/launch/gymkhana.launch (hunk 2 above)
Comment (mabelzhang) [reply vorc_gazebo/launch/gymkhana.launch]: Yes oops, good catch 0b9705c

File: vorc_gazebo/include/vorc_gazebo/gymkhana_scoring_plugin.hh (hunk 1 above)
Comment (mabelzhang) [reply vorc_gazebo/include/vorc_gazebo/gymkhana_scoring_plugin.hh:43]: 0b9705c

**Comment by mabelzhang** [comment]
Comments should have been addressed.
Wiki has been updated with topics of the task https://github.com/osrf/vorc/wiki/Tasks#4-black-box-gymkhana-task

**Review by caguero — APPROVED** [review]
Looks good!

**Pull request merged**